### PR TITLE
Add dev job to regenerate catalog from metadata and view diff

### LINKED ISSUE #24516: Add a job that parses metadata files into their catalogs
1. Take existing metadata files in dagster
2. Recompile them into the catalogs
3. diff the new catalogs with the old to ensure they match

### PR DESCRIPTION
**Pull request by bnchrch**
## What
image

closes #24516 

We need a way to see if the catalog that is generated by the new system matches the old system.

## How
1. Add the code to generate a catalog from metadata files
2. Adds a few dev assets that diff the old catalog and the new catalog and display the results

### CODE REVIEW

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:70]: nit to avoid an extra iteration.
```suggestion
    metadata_sources = [metadata for metadata in metadata_definitions if metadata["data"]["connectorType"] == "source" and metadata["data"]["catalogs"][catalog_name]["enabled"]
    metadata_destinations = [metadata for metadata in metadata_definitions if metadata["data"]["connectorType"] == "destination" and metadata["data"]["catalogs"][catalog_name]["enabled"]]
```

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:138]:
```suggestion
def key_catalog_entries(catalog: dict) -> dict:
```

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:140]:
```suggestion
    for connector_type, id_field in [("sources", "sourceDefinitionId"), ("destinations", "destinationDefinitionId")]:
```

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:11]: What do you think of keeping helpers in a separate module / package ?

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:230]: Could you share why these assets return dataframe and not pure DeepDiff objects?

**Review by alafanechere — COMMENTED** [review]
I don't know to what extent this code is temporary but I think the helpers you defined would greatly benefits from more typed models instead of pure dict interactions. I would find it more readable and maintainable to have a Metadata model with methods to transform it to desired intermediate models  or dict.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:230]: Good question: I have it return a dataframe to take advantage of the `to_markdown` method for previews

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:22]: I was definitely lazy with these as its meant to be short lived.

However lets move this to a dataclass like you suggest. The script is short lived but the translation from metadata to catalog is not

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:18]: Good call! And I appreciate all the python nits. Still getting back into the language!

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:16]: Great call out. To be honest I wasnt happy I was using a mutating fn here either. Will change.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:11]: Hmm good question, and generally im totally for helper/util files/modules/classes.

We already have the helper file pattern in this orchestration project as proof 😅 

However in my mind I see helper functions as filling two possible roles
1. Providing a piece of reusable code
2. Providing a way to self document code and break up the cognitive load when reading a function that uses the helper. (code should read like a book principle) 

If a helper functions only purpose is item #2 (self documentation) but we anticipate it will only ever be used in one circumstance by one other function/class/file, then I typically think it should be collocated. in the same file.

So in this instance I would opt to leave these specific helpers here. 

Does that change your mind any?

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py:1]: Fair call! Will change

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:230]: Would it make sense to provide a final asset only exposing the dataframe and depending on the oss_catalog_diff which returns the diff object? Might be overkill at the moment.

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/dev_assets.py:230]: It may slightly violate the "only write the code when you need to" rule 😛 

but its a good idea and what dagster is meant for so lets do it!

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_assets.py]: @alafanechere I added a PartialMetadataDefinition. Let me know what you think about the pattern.

Im going to hold off on adding a catalog definition model, as I wait for a decision on the airbyte-types repo at the end of this week.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:19]: Nit: as you reuse this *deepcopy* pattern you could implement a [decorator](https://realpython.com/primer-on-python-decorators/) to always deepcopy the input dict

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:84]:
```suggestion
def construct_catalog_from_metadata(catalog_derived_metadata_definitions: List[PartialMetadataDefinition], catalog_name: str) -> dict:
```

I'm not sure if you're passing a list of dict or PartialMetadataDefinition because `cloud_catalog_from_metadata` passes its `catalog_derived_metadata_definitions` input list composed of PartialMetadataDefinition.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:127]: nit: I think a custom decorator can be created and use  as you consistently return the same types in the same way on each assets below.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:8]: nit 
```suggestion
from metadata_service.orchestrator.utils.dagster_helpers import OutputDataFrame, output_dataframe
```

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:133]:
```suggestion
        return PartialMetadataDefinition(**metadata)
```
Just wondering why we're not building the model directly

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:98]: I think a docstring would be helpful here.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:102]:
```suggestion
sanitized_connectors = merged_connectors.fillna(None)
```

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:104]:
```suggestion
    def build_metadata(connector_catalog_entry: dict) -> PartialMetadataDefinition:
```

I found it hard to understand what was happening in this function until I realized you're manipulating a dict and not a dataframe will all the catalog entry.
Please add a docstring to explain what the format of the expected input and where it comes from.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:133]: Oh I think its because you explicitly want to delay the validation in a separate step.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:152]: I would expect this asset to return a list of valid `ConnectorMetadataDefinitionV1`.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py:8]: I find it's a breaking a bit the pydantic purpose if you delay to model validation.
I found an interesting suggestion [here](https://github.com/pydantic/pydantic/issues/668#issuecomment-512525178). Let me know what you think.

**Review by alafanechere — COMMENTED** [review]
Thank you for the use of a pydantic model 🙏 .
While reading the code I found it hard to figure what types you were actually handling in the functions due to some typehints / variable naming inconsistencies.
I would suggest to only use dataframe to merge oss and cloud catalog. And then serialize dataframe to tuple of `(ExpectedPydanticModel`, list_of_dict)` (a bit like what's suggested [here](https://github.com/pydantic/pydantic/issues/668#issuecomment-512525178)). And then reconvert to dataframe for the "view", the edge of your DAG. But maybe there's a dagster constraint I did not grasp.
I would be fine keep dataframe all along the way if you were always manipulating dataframe on which you apply pandas functions. You could also perform dataframe validation with pydantic with [this kind of approach.](https://www.inwt-statistics.com/read-blog/pandas-dataframe-validation-with-pydantic-part-2.html)

I'm fine approving this PR if you want to mainline it quickly but I'd love that we figure together a logic of when to use dict / list of dict / list of pydantic models / dataframe.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:108]: Thank you! Installing

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:127]: I think in this case, because we're already under the weight of dagster decorators I would like to keep whats happening for each asset much more explicit. 

My worry is two layers of decorators, is two layers of indirection, which can make debugging and understanding a little more difficult.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:133]: Yup exactly that, these aren't "valid" until overrides are applied

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py:152]: I can rename this!

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py:8]: @alafanechere Im glad you noticed this!

Its definately an awkward duck because we
1. Want strong type guarantees of what is metadata and what is not
2. Want these types to originate (be generated) from a jsonschema file to support java and python types
3. Want partial versions of these types to allow us to delay validation while we apply overrides

I went through a whole bunch of libraries / discussions on it.

Generally
1. Pydantic doesnt support partials out of the box
2. Pydantic doesnt seem to want to support this use case
3. The `pydantic-partial` library doesn't support nested validations 😢 

Ok so looking at the solution you linked. I like it, but I think theres a few small issues
1. Out of the box it doesn't give us access to the subscriptable methods __getitem__ and __setitem__
2. I actually tried to get this working but unfortunately theres its also not as easy as adding the subscriptable methods and we need to call some more internal apis to properly update the internal raw data dict (see screenshot below)
3. When looking at implementing this there were enough bugs I had to patch over that make me nervous to take at face value

In general I think im at the point where im happy to stay with the current implementation if it works. So long as you are 😅 

image


### Appendix
https://github.com/team23/pydantic-partial
https://github.com/pydantic/pydantic/issues/1673
https://github.com/tiangolo/fastapi/issues/125

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata.py]: Had to go a slightly different way but thanks for the prompt

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/models/metadata.py:8]: Fair enough, let's keep current implementation!

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:145]: Fair enough 😄

**Review by alafanechere — APPROVED** [review]
Thank you for your changes! I'm ok with the tradeoffs you made. I think you still want to finish the tests as you commented out some lines in these files.

File: airbyte-ci/connectors/metadata_service/orchestrator/tests/test_debug.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/tests/test_debug.py:64]: Ahh yes.

Sorry this is a little bit of jank introduced by Dagster.

See this comment at the top of the file:
https://github.com/airbytehq/airbyte/pull/24545/files/823e1f59eafe45dbabbbbabbe3316b2a0ee3fffd#diff-350aa46358bd8f37293007092536a25fb77f278e282f0b18ed61ab5d49053839L23-L28

File: airbyte-ci/connectors/metadata_service/orchestrator/tests/test_catalog.py
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/tests/test_catalog.py:54]: @alafanechere  its actually a comment!

I can see the mix up as it starts with `assert`

**Pull request merged**
